File: app/backend/java/payment-service/src/main/java/com/srbms/paymentservice/strategy/WalletPaymentProcessor.java
package com.srbms.paymentservice.strategy;

import com.srbms.paymentservice.dto.PaymentRequestDTO;
import com.srbms.paymentservice.entity.*;
import com.srbms.paymentservice.exception.InsufficientFundsException;
import com.srbms.paymentservice.repository.WalletRepository;
import com.srbms.paymentservice.repository.WalletTransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.UUID;

@Component
public class WalletPaymentProcessor implements PaymentProcessorStrategy {

    private final WalletRepository walletRepository;
    private final WalletTransactionRepository walletTransactionRepository;

    @Autowired
    public WalletPaymentProcessor(WalletRepository walletRepository,
                                  WalletTransactionRepository walletTransactionRepository) {
        this.walletRepository = walletRepository;
        this.walletTransactionRepository = walletTransactionRepository;
    }

    @Override
    public boolean supports(PaymentMode paymentMode) {
        return paymentMode == PaymentMode.WALLET;
    }

    @Override
    public Payment process(Payment payment, PaymentRequestDTO requestDTO) {
        Integer userId = payment.getUserId();
        Wallet wallet = walletRepository.findByUserId(userId)
                .orElseGet(() -> walletRepository.save(new Wallet(userId, BigDecimal.ZERO)));

        BigDecimal amountToPay = payment.getNetAmount();
        if (wallet.getBalance().compareTo(amountToPay) < 0) {
            payment.setPaymentStatus(PaymentStatus.FAILED);
            payment.setFailureReason("Insufficient wallet balance. Available: " + wallet.getBalance() + ", Required: " + amountToPay);
            throw new InsufficientFundsException(payment.getFailureReason());
        }

        // Deduct balance
        wallet.setBalance(wallet.getBalance().subtract(amountToPay));
        walletRepository.save(wallet);

        // Record transaction
        String refId = "WLT_DEBIT_" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        walletTransactionRepository.save(new WalletTransaction(
                wallet.getWalletId(),
                "DEBIT",
                amountToPay,
                refId,
                "Ride Payment for Ride #" + payment.getRideId()
        ));

        payment.setPaymentMode(PaymentMode.WALLET);
        payment.setGatewayRef(refId);
        payment.setPaymentStatus(PaymentStatus.SUCCESS);
        return payment;
    }

    @Override
    public Payment refund(Payment payment, BigDecimal amount, String reason) {
        Integer userId = payment.getUserId();
        Wallet wallet = walletRepository.findByUserId(userId)
                .orElseGet(() -> walletRepository.save(new Wallet(userId, BigDecimal.ZERO)));

        BigDecimal refundAmt = amount != null ? amount : payment.getNetAmount();
        wallet.setBalance(wallet.getBalance().add(refundAmt));
        walletRepository.save(wallet);

        String refId = "WLT_CREDIT_" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        walletTransactionRepository.save(new WalletTransaction(
                wallet.getWalletId(),
                "CREDIT",
                refundAmt,
                refId,
                "Refund for Ride #" + payment.getRideId() + ": " + reason
        ));

        payment.setPaymentStatus(PaymentStatus.REFUNDED);
        payment.setGatewayRef(refId);
        return payment;
    }
}

File: app/backend/.net/SmartRide.PaymentService/Strategies/WalletPaymentProcessor.cs
using Microsoft.EntityFrameworkCore;
using SmartRide.PaymentService.Data;
using SmartRide.PaymentService.Exceptions;
using SmartRide.PaymentService.Models.DTOs;
using SmartRide.PaymentService.Models.Entities;
using SmartRide.PaymentService.Models.Enums;

namespace SmartRide.PaymentService.Strategies;

public class WalletPaymentProcessor : IPaymentProcessorStrategy
{
    private readonly PaymentDbContext _dbContext;

    public WalletPaymentProcessor(PaymentDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public bool Supports(PaymentMode paymentMode)
    {
        return paymentMode == PaymentMode.WALLET;
    }

    public async Task<Payment> ProcessAsync(Payment payment, PaymentRequestDTO requestDTO)
    {
        int userId = payment.UserId ?? 0;
        var wallet = await _dbContext.Wallets.FirstOrDefaultAsync(w => w.UserId == userId);

        if (wallet == null)
        {
            wallet = new Wallet(userId, 0m);
            _dbContext.Wallets.Add(wallet);
            await _dbContext.SaveChangesAsync();
        }

        decimal amountToPay = payment.NetAmount;
        if (wallet.Balance < amountToPay)
        {
            payment.PaymentStatus = PaymentStatus.FAILED;
            payment.FailureReason = $"Insufficient wallet balance. Available: {wallet.Balance}, Required: {amountToPay}";
            throw new InsufficientFundsException(payment.FailureReason);
        }

        // Deduct balance
        wallet.Balance -= amountToPay;
        wallet.UpdatedAt = DateTime.UtcNow;

        // Record transaction
        string refId = "WLT_DEBIT_" + Guid.NewGuid().ToString("N")[..8].ToUpper();
        var walletTxn = new WalletTransaction(
            wallet.WalletId,
            "DEBIT",
            amountToPay,
            refId,
            $"Ride Payment for Ride #{payment.RideId}"
        );
        _dbContext.WalletTransactions.Add(walletTxn);

        await _dbContext.SaveChangesAsync();

        payment.PaymentMode = PaymentMode.WALLET;
        payment.GatewayRef = refId;
        payment.PaymentStatus = PaymentStatus.SUCCESS;
        return payment;
    }

    public async Task<Payment> RefundAsync(Payment payment, decimal? amount, string reason)
    {
        int userId = payment.UserId ?? 0;
        var wallet = await _dbContext.Wallets.FirstOrDefaultAsync(w => w.UserId == userId);

        if (wallet == null)
        {
            wallet = new Wallet(userId, 0m);
            _dbContext.Wallets.Add(wallet);
            await _dbContext.SaveChangesAsync();
        }

        decimal refundAmt = amount ?? payment.NetAmount;
        wallet.Balance += refundAmt;
        wallet.UpdatedAt = DateTime.UtcNow;

        string refId = "WLT_CREDIT_" + Guid.NewGuid().ToString("N")[..8].ToUpper();
        var walletTxn = new WalletTransaction(
            wallet.WalletId,
            "CREDIT",
            refundAmt,
            refId,
            $"Refund for Ride #{payment.RideId}: {reason}"
        );
        _dbContext.WalletTransactions.Add(walletTxn);

        await _dbContext.SaveChangesAsync();

        payment.PaymentStatus = PaymentStatus.REFUNDED;
        payment.GatewayRef = refId;
        return payment;
    }
}
